File: src/models/record.ts
import { Document, model, Schema } from 'mongoose';
import pick from 'lodash/pick';
import uuidv4 from 'uuid/v4';

import { Record } from '../types';
import { RecordStatus } from '../types/enums';

export interface RecordDocument extends Document, Record {}

const recordSchemaDefinition = {
  id: {
    default: uuidv4,
    type: String,
    required: true,
    unique: true
  },
  organizationId: {
    type: String,
    ref: 'organization'
  },
  status: { type: String, default: RecordStatus.DRAFT },

  // Behandlingen gjelder
  title: { type: String },
  // Formålet med behandlingen
  purpose: { type: String },
  // Kategorier av personopplysninger og registrerte
  categories: [
    {
      dataSubjectCategories: { type: String },
      personalDataCategories: [String]
    }
  ],
  dataSubjectCategories: String,
  // Behandlingsgrunnlagt artikkel 6
  articleSixBasis: {
    type: [
      {
        _id: false,
        legality: String,
        referenceUrl: String
      }
    ],
    default: null
  },
  // Behandlingsgrunnlag artikkel 9 eller 10
  otherArticles: {
    articleNine: {
      checked: Boolean,
      referenceUrl: String
    },
    articleTen: {
      checked: Boolean,
      referenceUrl: String
    }
  },

  // Funksjonsområde
  businessAreas: [String],
  // Tilhørende datasett
  relatedDatasets: [String],

  // Databehandlere og databehandleravtaler
  dataProcessingAgreements: {
    type: [
      {
        _id: false,
        dataProcessorName: { type: String },
        agreementUrl: { type: String }
      }
    ]
  },
  // Daglig behandlingsansvarlig
  dataProcessorContactDetails: {
    type: [
      {
        _id: false,
        name: { type: String },
        email: { type: String },
        phone: { type: String }
      }
    ]
  },
  // Felles behandlingsansvarlig
  commonDataControllerContact: {
    type: {
      companies: String,
      distributionOfResponsibilities: String,
      contactPoints: {
        type: [
          {
            name: String,
            email: String,
            phone: String
          }
        ]
      }
    }
  },
  // Beskrivelse av tekniske og organisatoriske sikkerhetstiltak
  securityMeasures: String,
  // Planlagte tidsfrister for sletting
  plannedDeletion: String,
  // Høy personvernrisiko
  highPrivacyRisk: Boolean,
  // Risikovurdering/DPIA gjennomført
  dataProtectionImpactAssessment: {
    conducted: Boolean,
    assessmentReportUrl: String
  },
  // Kilder til personopplysninger
  personalDataSubjects: String,
  // sist oppdatert
  updatedAt: Date,
  // Systemer i din virksomhet som behandler personopplysningene
  privacyProcessingSystems: String,
  // Kategorier av mottakere
  recipientCategories: [String],
  // Overføres opplysninger til tredjeland?
  dataTransfers: {
    transferred: Boolean,
    thirdCountryRecipients: String,
    guarantees: String
  }
};

const recordFromDocument = (document: RecordDocument): Record =>
  pick(document, Object.keys(recordSchemaDefinition));

const RecordSchema = new Schema(recordSchemaDefinition, {
  timestamps: true,
  strict: true,
  toObject: {
    transform: recordFromDocument
  }
});

export const RecordModel = model<RecordDocument>('record', RecordSchema);

File: src/main/kotlin/no/fdk/records_of_processing_activities/model/Record.kt
package no.fdk.records_of_processing_activities.model

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonInclude
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.mapping.Document
import java.time.LocalDateTime

@Document(collection = "records")
data class RecordDBO(
    @Id
    val id: String,
    val status: RecordStatus?,
    val organizationId: String?,
    val purpose: String?,
    val title: String?,
    val categories: ArrayList<Categories>?,
    val articleSixBasis: ArrayList<ArticleSix>?,
    val otherArticles: OtherArticles?,
    val businessAreas: ArrayList<String>?,
    val relatedDatasets: ArrayList<String>?,
    val dataProcessingAgreements: ArrayList<DataProcessingAgreement>?,
    val dataProcessorContactDetails: ArrayList<ContactPoint>?,
    val commonDataControllerContact: CommonDataControllerContact?,
    val personalDataSubjects: String?,
    val securityMeasures: String?,
    val plannedDeletion: String?,
    val dataProtectionImpactAssessment: DataProtectionImpactAssessment?,
    val privacyProcessingSystems: String?,
    val recipientCategories: ArrayList<String>?,
    val dataTransfers: DataTransfers?,
    val updatedAt: LocalDateTime?
)

@JsonIgnoreProperties(ignoreUnknown = true)
@JsonInclude(JsonInclude.Include.NON_EMPTY)
data class RecordDTO(
    val id: String?,
    val status: RecordStatus?,
    val organizationId: String?,
    val purpose: String?,
    val title: String?,
    val categories: ArrayList<Categories>?,
    val articleSixBasis: ArrayList<ArticleSix>?,
    val otherArticles: OtherArticles?,
    val businessAreas: ArrayList<String>?,
    val relatedDatasets: ArrayList<String>?,
    val dataProcessingAgreements: ArrayList<DataProcessingAgreement>?,
    val dataProcessorContactDetails: ArrayList<ContactPoint>?,
    val commonDataControllerContact: CommonDataControllerContact?,
    val personalDataSubjects: String?,
    val securityMeasures: String?,
    val plannedDeletion: String?,
    val dataProtectionImpactAssessment: DataProtectionImpactAssessment?,
    val privacyProcessingSystems: String?,
    val recipientCategories: ArrayList<String>?,
    val dataTransfers: DataTransfers?,
    val updatedAt: LocalDateTime?
)

data class PagedRecords(
    val pageNumber: Int,
    val pagesTotal: Int,
    val size: Int,
    val hits: List<RecordDTO>
)

data class ContactPoint(
    val name: String?,
    val email: String?,
    val phone: String?
)

data class Categories(
    val dataSubjectCategories: String?,
    val personalDataCategories: ArrayList<String>?
)

data class Article(
    val checked: Boolean?,
    val referenceUrl: String?
)

data class ArticleSix(
    val legality: String?,
    val referenceUrl: String?
)

data class ArticleLegality(
    val legality: String?,
    val checked: Boolean?,
    val referenceUrl: String?
)

data class ArticleWithLegalities(
    val checked: Boolean?,
    val referenceUrl: String?,
    val legalities: ArrayList<ArticleLegality>?
)

data class OtherArticles(
    val articleNine: ArticleWithLegalities?,
    val articleTen: Article?
)

data class DataProcessingAgreement(
    val dataProcessorName: String?,
    val agreementUrl: String?
)

data class CommonDataControllerContact(
    val commonDataControllerChecked: Boolean?,
    val companies: String?,
    val distributionOfResponsibilities: String?,
    val contactPoints: ArrayList<ContactPoint>?
)

data class DataProtectionImpactAssessment(
    val conducted: Boolean?,
    val assessmentReportUrl: String?
)

data class DataTransfers(
    val transferred: Boolean?,
    val thirdCountryRecipients: String?,
    val guarantees: String?,
)

enum class RecordStatus {
    DRAFT, APPROVED
}
